File: src/infrastructure/environment/sandbox.py
"""Sandbox abstractions for running tool commands in isolated environments.

- NullSandbox: runs commands directly on the host. Current Phase 1 behaviour.
  Kept for dev laptops without Docker and for tests.
- DockerSandbox: runs commands inside a disposable container with a mounted
  workspace, no network, cpu/mem caps, non-root user.

Availability is detected once per process. Commands that the policy calls
through `run_command` flow through whichever sandbox the RepoEnvironmentRunner
is configured with; file reads/lists/searches stay on the host (they're
read-only against a path-checked workspace).
"""
from __future__ import annotations

import shutil
import subprocess
from dataclasses import dataclass
from pathlib import Path
from typing import Protocol


@dataclass(frozen=True)
class SandboxResult:
    returncode: int
    stdout: str
    stderr: str
    timed_out: bool = False


class Sandbox(Protocol):
    name: str

    def available(self) -> bool: ...

    def run(
        self,
        workspace: Path,
        command: list[str],
        *,
        timeout: float,
    ) -> SandboxResult: ...


@dataclass(frozen=True)
class NullSandbox:
    """Host-side subprocess. No isolation beyond the path/command allowlist."""

    name: str = "none"

    def available(self) -> bool:
        return True

    def run(self, workspace: Path, command: list[str], *, timeout: float) -> SandboxResult:
        try:
            out = subprocess.run(
                command, cwd=workspace, capture_output=True, text=True,
                timeout=timeout, check=False,
            )
        except subprocess.TimeoutExpired as exc:
            return SandboxResult(
                returncode=124,
                stdout=(exc.stdout or b"").decode("utf-8", errors="replace") if isinstance(exc.stdout, bytes) else (exc.stdout or ""),
                stderr="timeout",
                timed_out=True,
            )
        return SandboxResult(returncode=out.returncode, stdout=out.stdout, stderr=out.stderr)


@dataclass
class DockerSandbox:
    """Runs commands inside a disposable container with workspace mounted at /ws."""

    image: str = "python:3.11-slim"
    memory: str = "512m"
    cpus: str = "1"
    network: str = "none"
    user: str = "nobody"
    name: str = "docker"
    _available: bool | None = None

    def available(self) -> bool:
        if self._available is not None:
            return self._available
        if shutil.which("docker") is None:
            self._available = False
            return False
        try:
            subprocess.run(
                ["docker", "version", "--format", "{{.Server.Version}}"],
                capture_output=True, check=True, timeout=3,
            )
        except (subprocess.CalledProcessError, subprocess.TimeoutExpired, FileNotFoundError):
            self._available = False
            return False
        self._available = True
        return True

    def run(self, workspace: Path, command: list[str], *, timeout: float) -> SandboxResult:
        docker_cmd = [
            "docker", "run", "--rm",
            f"--network={self.network}",
            f"--memory={self.memory}",
            f"--cpus={self.cpus}",
            "--user", self.user,
            "-v", f"{workspace}:/ws",
            "-w", "/ws",
            self.image,
            *command,
        ]
        try:
            out = subprocess.run(
                docker_cmd, capture_output=True, text=True,
                timeout=timeout + 5, check=False,
            )
        except subprocess.TimeoutExpired:
            return SandboxResult(returncode=124, stdout="", stderr="timeout", timed_out=True)
        return SandboxResult(returncode=out.returncode, stdout=out.stdout, stderr=out.stderr)


def build_sandbox(kind: str, image: str | None = None) -> Sandbox:
    """Factory used by bootstrap. Falls back to NullSandbox if docker unavailable."""
    if kind == "none":
        return NullSandbox()
    if kind == "docker":
        sb = DockerSandbox(image=image) if image else DockerSandbox()
        if sb.available():
            return sb
        return NullSandbox()
    raise ValueError(f"Unknown sandbox kind: {kind}")

File: crates/horizon-runner/src/sandbox.rs
//! Sandboxes for tool commands.
//!
//! `SandboxKind::None` runs on the host with only the allowlist between
//! the policy and the machine; keep it for laptops and tests.
//! `SandboxKind::Docker` runs each command in a disposable container:
//! no network, cpu and memory caps, non-root, workspace mounted at /ws.

use std::path::Path;
use std::process::Stdio;
use std::time::Duration;

use tokio::process::Command;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SandboxResult {
    pub returncode: i32,
    pub stdout: String,
    pub stderr: String,
    pub timed_out: bool,
}

#[derive(Debug, Clone)]
pub enum SandboxKind {
    None,
    Docker {
        image: String,
        memory: String,
        cpus: String,
        network: String,
        user: String,
    },
}

impl SandboxKind {
    pub fn docker(image: Option<&str>) -> Self {
        SandboxKind::Docker {
            image: image.unwrap_or("python:3.11-slim").to_string(),
            memory: "512m".into(),
            cpus: "1".into(),
            network: "none".into(),
            user: "nobody".into(),
        }
    }
}

#[derive(Debug, Clone)]
pub struct Sandbox {
    kind: SandboxKind,
}

impl Sandbox {
    pub fn new(kind: SandboxKind) -> Self {
        Self { kind }
    }

    pub fn host() -> Self {
        Self::new(SandboxKind::None)
    }

    /// Build from a settings string. Docker falls back to the host sandbox
    /// when the daemon is unreachable, mirroring the previous behaviour.
    pub async fn from_setting(kind: &str, image: Option<&str>) -> Result<Self, String> {
        match kind {
            "none" => Ok(Self::host()),
            "docker" => {
                if docker_available().await {
                    Ok(Self::new(SandboxKind::docker(image)))
                } else {
                    tracing::warn!("docker unavailable; falling back to host sandbox");
                    Ok(Self::host())
                }
            }
            other => Err(format!("Unknown sandbox kind: {other}")),
        }
    }

    pub fn name(&self) -> &'static str {
        match self.kind {
            SandboxKind::None => "none",
            SandboxKind::Docker { .. } => "docker",
        }
    }

    pub async fn run(&self, workspace: &Path, argv: &[String], timeout: Duration) -> SandboxResult {
        let (program, args, budget): (String, Vec<String>, Duration) = match &self.kind {
            SandboxKind::None => (argv[0].clone(), argv[1..].to_vec(), timeout),
            SandboxKind::Docker {
                image,
                memory,
                cpus,
                network,
                user,
            } => {
                let mut args = vec![
                    "run".to_string(),
                    "--rm".into(),
                    format!("--network={network}"),
                    format!("--memory={memory}"),
                    format!("--cpus={cpus}"),
                    "--user".into(),
                    user.clone(),
                    "-v".into(),
                    format!("{}:/ws", workspace.display()),
                    "-w".into(),
                    "/ws".into(),
                    image.clone(),
                ];
                args.extend(argv.iter().cloned());
                ("docker".into(), args, timeout + Duration::from_secs(5))
            }
        };
        let mut cmd = Command::new(&program);
        cmd.args(&args)
            .current_dir(workspace)
            .stdin(Stdio::null())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .kill_on_drop(true);
        let child = match cmd.spawn() {
            Ok(c) => c,
            Err(e) => {
                return SandboxResult {
                    returncode: 127,
                    stdout: String::new(),
                    stderr: format!("spawn failed: {e}"),
                    timed_out: false,
                }
            }
        };
        match tokio::time::timeout(budget, child.wait_with_output()).await {
            Ok(Ok(out)) => SandboxResult {
                returncode: out.status.code().unwrap_or(-1),
                stdout: String::from_utf8_lossy(&out.stdout).into_owned(),
                stderr: String::from_utf8_lossy(&out.stderr).into_owned(),
                timed_out: false,
            },
            Ok(Err(e)) => SandboxResult {
                returncode: -1,
                stdout: String::new(),
                stderr: e.to_string(),
                timed_out: false,
            },
            // Dropping the future kills the child (kill_on_drop).
            Err(_) => SandboxResult {
                returncode: 124,
                stdout: String::new(),
                stderr: "timeout".into(),
                timed_out: true,
            },
        }
    }
}

async fn docker_available() -> bool {
    let probe = Command::new("docker")
        .args(["version", "--format", "{{.Server.Version}}"])
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .kill_on_drop(true)
        .status();
    matches!(tokio::time::timeout(Duration::from_secs(3), probe).await, Ok(Ok(s)) if s.success())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn host_sandbox_runs_and_times_out() {
        let dir = tempfile::tempdir().unwrap();
        let sb = Sandbox::host();
        let ok = sb
            .run(
                dir.path(),
                &["echo".into(), "hi".into()],
                Duration::from_secs(5),
            )
            .await;
        assert_eq!(ok.returncode, 0);
        assert_eq!(ok.stdout.trim(), "hi");
        let slow = sb
            .run(
                dir.path(),
                &["sleep".into(), "5".into()],
                Duration::from_millis(100),
            )
            .await;
        assert!(slow.timed_out);
        assert_eq!(slow.returncode, 124);
    }
}
